File: aifo/slideinsight_api/slideinsight_sdk/resources/slides.py
"""
Slides resource manager for SlideInsight SDK.

Provides operations for managing slides including creation, metadata retrieval,
tile access, and annotations management.
"""

from __future__ import annotations

from typing import Optional

from .base import BaseResourceManager
from ..models import (
    Slide,
    SlideMetadata,
    SlideTile,
    SlideAnnotationsOverview,
    RasterAnnotation,
    VectorAnnotation,
    PaginatedResponse,
    PaginationInfo,
)


class SlidesManager(BaseResourceManager[Slide]):
    """Manager for slide resources."""

    def __init__(self, http_client):
        super().__init__(
            http_client=http_client, base_endpoint="/api/v1/slides", model_class=Slide, list_response_key="slides"
        )

    async def create(
        self,
        slide_uri: str,
        slide_name: Optional[str] = None,
        slide_id: Optional[str] = None,
        case_uid: Optional[str] = None,
    ) -> Slide:
        """Create a new slide.

        Args:
            slide_uri: URI/path to the slide file
            slide_name: Human-readable slide name
            slide_id: Slide identifier
            case_uid: Case UID to associate with (optional)

        Returns:
            Created slide
        """
        data = {
            "slideUri": slide_uri,
        }

        if slide_name is not None:
            data["slideName"] = slide_name
        if slide_id is not None:
            data["slideId"] = slide_id
        if case_uid is not None:
            data["caseUid"] = case_uid

        response_data = await self.http_client.request("POST", self.base_endpoint, json=data)
        return Slide.from_dict(response_data)

    async def get_metadata(self, slide_uid: str) -> SlideMetadata:
        """Get slide metadata including dimensions and properties.

        Args:
            slide_uid: Slide UID

        Returns:
            Slide metadata
        """
        endpoint = self._build_endpoint(f"{slide_uid}/metadata")
        response_data = await self.http_client.request("GET", endpoint)
        return SlideMetadata.from_dict(response_data)

    async def get_tile(
        self,
        slide_uid: str,
        z: int,
        x: int,
        y: int,
        format: str = "jpeg",
    ) -> SlideTile:
        """Get a tile from a slide.

        Args:
            slide_uid: Slide UID
            z: Zoom level
            x: Tile X coordinate
            y: Tile Y coordinate
            format: Image format ('jpeg' or 'png')

        Returns:
            Slide tile with image data
        """
        endpoint = f"/api/v1/slides/{slide_uid}/tiles/{z}/{x}/{y}.{format}"
        content_type, image_data = await self.http_client.request_binary("GET", endpoint)
        return SlideTile.from_response(content_type, image_data)

    async def get_annotations_overview(self, slide_uid: str) -> SlideAnnotationsOverview:
        """Get overview of annotations for a slide.

        Args:
            slide_uid: Slide UID

        Returns:
            Annotations overview with counts and URLs
        """
        endpoint = self._build_endpoint(f"{slide_uid}/annotations")
        response_data = await self.http_client.request("GET", endpoint)
        return SlideAnnotationsOverview.from_dict(response_data)

    async def get_raster_annotations(
        self,
        slide_uid: str,
        page: int = 1,
        limit: int = 100,
    ) -> list[RasterAnnotation]:
        """Get raster annotations (masks) for a slide.

        Args:
            slide_uid: Slide UID
            page: Page number
            limit: Items per page

        Returns:
            List of raster annotations
        """
        endpoint = f"/api/v1/slides/{slide_uid}/annotations/raster"
        params = self._build_params(page=page, limit=limit)

        response_data = await self.http_client.request("GET", endpoint, params=params)

        # Handle response format (could be different structures)
        if "masks" in response_data:
            masks_data = response_data["masks"]
        elif "rasterAnnotations" in response_data:
            masks_data = response_data["rasterAnnotations"]
        else:
            masks_data = response_data.get("items", response_data.get("data", []))

        return [RasterAnnotation.from_dict(mask) for mask in masks_data]

    async def add_raster_annotation(
        self,
        slide_uid: str,
        mask_uri: str,
        mask_name: Optional[str] = None,
        actor_type: str = "user",
        actor_id: Optional[int] = None,
        labels: Optional[dict] = None,
        metadata: Optional[dict] = None,
    ) -> RasterAnnotation:
        """Add a raster annotation (mask) to a slide.

        Args:
            slide_uid: Slide UID
            mask_uri: URI/path to the mask file
            mask_name: Human-readable mask name
            actor_type: Type of actor ('user' or 'model')
            actor_id: ID of the actor
            labels: Annotation labels
            metadata: Additional metadata

        Returns:
            Created raster annotation
        """
        endpoint = f"/api/v1/slides/{slide_uid}/annotations/raster"
        data = {
            "maskUri": mask_uri,
            "slideUid": slide_uid,
            "actorType": actor_type,
        }

        if mask_name is not None:
            data["maskName"] = mask_name
        if actor_id is not None:
            data["actorId"] = actor_id
        if labels is not None:
            data["labels"] = labels
        if metadata is not None:
            data["metadata"] = metadata

        response_data = await self.http_client.request("POST", endpoint, json=data)
        return RasterAnnotation.from_dict(response_data)

    async def get_vector_annotations(
        self,
        slide_uid: str,
        page: int = 1,
        limit: int = 100,
    ) -> list[VectorAnnotation]:
        """Get vector annotations for a slide.

        Args:
            slide_uid: Slide UID
            page: Page number
            limit: Items per page

        Returns:
            List of vector annotations
        """
        endpoint = f"/api/v1/slides/{slide_uid}/annotations/vector"
        params = self._build_params(page=page, limit=limit)

        response_data = await self.http_client.request("GET", endpoint, params=params)

        # Handle response format
        if "vectorAnnotations" in response_data:
            annotations_data = response_data["vectorAnnotations"]
        else:
            annotations_data = response_data.get("items", response_data.get("data", []))

        return [VectorAnnotation.from_dict(annotation) for annotation in annotations_data]

    async def add_vector_annotation(
        self,
        slide_uid: str,
        file_uri: str,
        name: Optional[str] = None,
        format: str = "geojson",
        actor_type: str = "user",
        actor_id: Optional[int] = None,
        labels: Optional[dict] = None,
        metadata: Optional[dict] = None,
    ) -> VectorAnnotation:
        """Add a vector annotation to a slide.

        Args:
            slide_uid: Slide UID
            file_uri: URI/path to the annotation file
            name: Human-readable annotation name
            format: Annotation format ('geojson', 'protobuf', 'zarr')
            actor_type: Type of actor ('user' or 'model')
            actor_id: ID of the actor
            labels: Annotation labels
            metadata: Additional metadata

        Returns:
            Created vector annotation
        """
        endpoint = f"/api/v1/slides/{slide_uid}/annotations/vector"
        data = {
            "fileUri": file_uri,
            "slideUid": slide_uid,
            "format": format,
            "actorType": actor_type,
        }

        if name is not None:
            data["name"] = name
        if actor_id is not None:
            data["actorId"] = actor_id
        if labels is not None:
            data["labels"] = labels
        if metadata is not None:
            data["metadata"] = metadata

        response_data = await self.http_client.request("POST", endpoint, json=data)
        return VectorAnnotation.from_dict(response_data)

    async def get_raster_annotation_tile(
        self,
        slide_uid: str,
        mask_id: str,
        z: int,
        x: int,
        y: int,
        format: str = "png",
    ) -> SlideTile:
        """Get a tile from a raster annotation.

        Args:
            slide_uid: Slide UID
            mask_id: Mask/annotation ID
            z: Zoom level
            x: Tile X coordinate
            y: Tile Y coordinate
            format: Image format ('png' recommended for masks)

        Returns:
            Annotation tile with image data
        """
        endpoint = f"/api/v1/slides/{slide_uid}/annotations/raster/{mask_id}/tiles/{z}/{x}/{y}.{format}"
        content_type, image_data = await self.http_client.request_binary("GET", endpoint)
        return SlideTile.from_response(content_type, image_data)

    async def get_vector_annotation_file(
        self,
        slide_uid: str,
        vector_id: str,
    ) -> bytes:
        """Get the raw file data for a vector annotation.

        Args:
            slide_uid: Slide UID
            vector_id: Vector annotation ID

        Returns:
            Raw file data
        """
        endpoint = f"/api/v1/slides/{slide_uid}/annotations/vector/{vector_id}/file"
        content_type, file_data = await self.http_client.request_binary("GET", endpoint)
        return file_data

File: aifo/slideinsight/internal/server/handlers/slides.go
// Copyright 2025 Jonas Teuwen. All rights reserved.
//
// This file is part of SlideInsight.
//
// Use of this source code is governed by the terms found in the
// LICENSE file located in the SlideInsight project root.
package handlers

import (
	"errors"
	"strconv"

	"aifo.dev/aifo/slideinsight/internal/server/domain"
	apperrors "aifo.dev/aifo/slideinsight/internal/server/errors"
	"aifo.dev/aifo/slideinsight/internal/server/middleware"
	"aifo.dev/aifo/slideinsight/internal/server/services"
	"aifo.dev/aifo/slideinsight/internal/server/validation"
	"aifo.dev/aifo/slideinsight/internal/utils"
	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/fiber/v2/log"
)

// slidesResponseBuilder builds a SlidesResponse from slides and pagination info
func slidesResponseBuilder(slides []domain.Slide, pagination domain.PaginationInfo) domain.SlidesResponse {
	return domain.SlidesResponse{
		Slides:     slides,
		Pagination: pagination,
	}
}

// GetSlides returns a handler function that retrieves all slides with optional search/filter support
// @Summary Get all slides
// @Description Retrieve a list of slides with optional search/filter and pagination support
// @Tags slides
// @Produce json
// @Security BearerAuth
// @Param page query int false "Page number (default: 1)" minimum(1) example(1)
// @Param limit query int false "Items per page (default: 100)" minimum(1) maximum(1000) example(100)
// @Param q query string false "General search across slide name and URI" example("sample slide")
// @Param slide_name query string false "Filter by slide name" example("slide-001")
// @Param sort query string false "Sort field (slide_name, created_at, slide_id)" example("slide_name")
// @Param dir query string false "Sort direction (asc, desc)" example("asc")
// @Success 200 {object} domain.SlidesResponse "List of slides with pagination"
// @Failure 400 {object} domain.ErrorResponse "Bad request - invalid parameters"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides [get]
func GetSlides(service services.SlidesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		params, err := utils.ParsePaginationAndSearchParams(c)
		if err != nil {
			return err // The error already contains the proper status code and message
		}

		// Use search method which handles both search and non-search cases
		slides, paginationInfo, err := service.GetSlidesGeneric(c.UserContext(), params)
		if err != nil {
			log.Error("GetSlides failed", "error", err, "search", params.SearchParams)
			return middleware.HandleError(c, err)
		}

		return c.JSON(domain.SlidesResponse{
			Slides:     slides,
			Pagination: paginationInfo,
		})
	}
}

// GetSlidesCount returns a handler function that retrieves the total count of slides
// @Summary Get slides count
// @Description Retrieve the total count of slides in the system
// @Tags slides
// @Produce json
// @Security BearerAuth
// @Success 200 {object} fiber.Map "Total count of slides"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides/count [get]
func GetSlidesCount(service services.SlidesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		count, err := service.GetSlidesCount(c.UserContext())
		if err != nil {
			log.Error("GetSlidesCount failed", "error", err)
			return middleware.HandleError(c, err)
		}

		return c.JSON(fiber.Map{
			"count": count,
		})
	}
}

func GetSlidesByCaseUID(service services.SlidesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		caseUID := c.Params("caseUID")
		slides, err := service.GetSlidesByCaseUID(c.UserContext(), caseUID)
		if err != nil {
			log.Error("GetSlidesByCaseUID failed", "error", err)
			return middleware.HandleError(c, err)
		}

		return c.JSON(fiber.Map{
			"slides": slides,
		})
	}
}

// GetSlideByUID returns a handler function that retrieves a specific slide by its ID
// @Summary Get slide by ID
// @Description Retrieve a specific slide by its unique identifier
// @Tags slides
// @Produce json
// @Security BearerAuth
// @Param slideUid path string true "Slide UID" example("slide123")
// @Success 200 {object} domain.Slide "Slide information"
// @Failure 400 {object} domain.ErrorResponse "Bad request - slideUid required"
// @Failure 404 {object} domain.ErrorResponse "Slide not found"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides/{slideUid} [get]
func GetSlideByUID(service services.SlidesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var input SlideUIDParams
		if err := c.ParamsParser(&input); err != nil {
			log.Warn("GetSlideByUID request parsing failed", "error", err)
			return middleware.SendError(c, fiber.StatusBadRequest, "invalid request")
		}

		// Validate the parsed parameters
		if err := validation.GlobalValidator.ValidateStruct(c, input); err != nil {
			return err // Error already formatted and sent
		}

		slide, err := service.GetSlideByUID(c.UserContext(), input.SlideUID)
		if err != nil {
			log.Error("GetSlideByUID failed", "error", err, "slideUID", input.SlideUID)
			if errors.Is(err, apperrors.ErrSlideNotFound) {
				return middleware.SendError(c, fiber.StatusNotFound, "slide not found")
			}
			return middleware.HandleError(c, err)
		}

		return c.JSON(slide)
	}
}

// GetSlideMetadata returns a handler function that retrieves detailed metadata for a specific slide
// @Summary Get slide metadata
// @Description Retrieve detailed metadata and properties for a specific slide
// @Tags slides
// @Produce json
// @Security BearerAuth
// @Param slideUid path string true "Slide ID" example("slide123")
// @Success 200 {object} domain.SlideMetadata "Slide metadata"
// @Failure 400 {object} domain.ErrorResponse "Bad request - slideUid required"
// @Failure 404 {object} domain.ErrorResponse "Slide not found"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides/{slideUid}/metadata [get]
func GetSlideMetadata(service services.SlidesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var input SlideUIDParams
		if err := c.ParamsParser(&input); err != nil {
			log.Warn("GetSlideMetadata request parsing failed", "error", err)
			return middleware.SendError(c, fiber.StatusBadRequest, "invalid request")
		}

		// Validate the parsed parameters
		if err := validation.GlobalValidator.ValidateStruct(c, input); err != nil {
			return err // Error already formatted and sent
		}

		metadata, err := service.GetSlideMetadata(c.UserContext(), input.SlideUID)
		if err != nil {
			log.Error("GetSlideMetadata failed", "error", err, "slideUid", input.SlideUID)
			if errors.Is(err, apperrors.ErrSlideNotFound) {
				return middleware.SendError(c, fiber.StatusNotFound, "slide not found")
			}
			return middleware.HandleError(c, err)
		}

		return c.JSON(metadata)
	}
}

// GetSlideTile handles requests for slide tiles
// @Summary Get slide tile
// @Description Retrieve a specific tile from a slide in XYZ tile format, or JXL (JPEG XL) format
// @Tags slides
// @Produce image/jpeg,image/png,image/jxl,application/octet-stream
// @Security BearerAuth
// @Param slideUid path string true "Slide ID" example("slide123")
// @Param z path int true "Zoom level" example(0)
// @Param x path int true "Tile X coordinate" example(0)
// @Param y path int true "Tile Y coordinate" example(0)
// @Param format path string true "Image format" Enums(jpg, png, jxl) example("jpg")
// @Param q query int false "Quality (1-100). Supported for JPEG, and JXL formats. PNG format will return error if set to non-100 value. Default is 75 for JXL, and JPEG." minimum(1) maximum(100) example(75)
// @Success 200 {file} binary "Tile image data, or JXL compressed data"
// @Failure 400 {object} domain.ErrorResponse "Bad request - invalid parameters"
// @Failure 404 {object} domain.ErrorResponse "Slide or tile not found"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides/{slideUid}/tiles/{z}/{x}/{y}.{format} [get]
func GetSlideTile(service services.SlidesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var input SlideTileParams
		if err := c.ParamsParser(&input); err != nil {
			log.Warn("GetSlideTile request parsing failed", "error", err)
			return middleware.SendError(c, fiber.StatusBadRequest, "invalid request")
		}

		// Manually parse quality query parameter since c.ParamsParser() only handles path params
		qualityStr := c.Query("q")
		if qualityStr != "" {
			if quality, err := strconv.Atoi(qualityStr); err == nil {
				if quality >= 1 && quality <= 100 {
					input.Quality = quality
				} else {
					log.Warn("Invalid quality parameter value", "quality", quality, "slideUID", input.SlideUID)
					return middleware.SendError(c, fiber.StatusBadRequest, "quality parameter must be between 1 and 100")
				}
			} else {
				log.Warn("Invalid quality parameter format", "qualityStr", qualityStr, "slideUID", input.SlideUID)
				return middleware.SendError(c, fiber.StatusBadRequest, "quality parameter must be a valid integer")
			}
		}

		// Validate the parsed parameters
		if err := validation.GlobalValidator.ValidateStruct(c, input); err != nil {
			return err // Error already formatted and sent
		}

		// Get the tile
		tile, err := service.GetSlideTile(c.UserContext(), input.SlideUID, input.Z, input.X, input.Y, input.Format, input.Quality)
		if err != nil {
			log.Error("GetSlideTile failed", "error", err, "slideUid", input.SlideUID, "z", input.Z, "x", input.X, "y", input.Y)
			if errors.Is(err, apperrors.ErrSlideNotFound) {
				return middleware.SendError(c, fiber.StatusNotFound, "slide not found")
			}
			if errors.Is(err, apperrors.ErrInvalidFormat) {
				return middleware.SendError(c, fiber.StatusBadRequest, err.Error())
			}
			if apperrors.IsOutOfBounds(err) {
				return middleware.SendError(c, fiber.StatusNotFound, "tile not found")
			}
			return middleware.HandleError(c, err)
		}

		// Set the content type header explicitly
		if input.Format == "jpg" {
			c.Set("Content-Type", "image/jpeg")
		} else if input.Format == "png" {
			c.Set("Content-Type", "image/png")
		} else if input.Format == "jxl" {
			c.Set("Content-Type", "application/octet-stream")
		}

		// Set status code explicitly
		c.Status(fiber.StatusOK)

		return c.Send(tile.Image)
	}
}

// AddSlide returns a handler function that adds a slide
// @Summary Add a new slide
// @Description Add a new slide to the system
// @Tags slides
// @Accept json
// @Produce json
// @Security BearerAuth
// @Param slide body SlideCreationInput true "Slide information"
// @Success 201 {object} domain.Slide "Created slide"
// @Failure 400 {object} domain.ErrorResponse "Bad request - invalid input"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides [post]
func AddSlide(service services.SlidesService, imageTypesService services.ImageTypesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var input SlideCreationInput
		if err := c.BodyParser(&input); err != nil {
			log.Warn("AddSlide request parsing failed", "error", err)
			return middleware.SendError(c, fiber.StatusBadRequest, "invalid request")
		}

		// Validate
		if input.SlideURI == "" {
			return middleware.SendError(c, fiber.StatusBadRequest, "slideUri is required")
		}

		// Validate image type ID if provided
		if input.ImageTypeId != "" {
			_, err := imageTypesService.GetImageTypeByID(c.UserContext(), input.ImageTypeId)
			if err != nil {
				log.Warn("Invalid image type ID provided", "imageTypeId", input.ImageTypeId, "error", err)
				return middleware.SendError(c, fiber.StatusBadRequest, "invalid image type ID")
			}
		}

		slide := domain.Slide{
			SlideUID:    input.SlideUID,
			SlideName:   input.SlideName,
			SlideURI:    input.SlideURI,
			ImageTypeId: input.ImageTypeId,
		}

		createdSlide, err := service.SaveSlide(c.UserContext(), slide)
		if err != nil {
			log.Error("AddSlide failed", "error", err)
			return middleware.HandleError(c, err)
		}

		// Return 201 Created with the slide data
		c.Status(fiber.StatusCreated)
		return c.JSON(createdSlide)
	}
}

// GetSlideAnnotationsOverview returns a handler function that provides an overview of annotations for a slide
// @Summary Get slide annotations overview
// @Description Get URLs and counts for both raster and vector annotations for a specific slide
// @Tags slides
// @Produce json
// @Security BearerAuth
// @Param slideUid path string true "Slide ID" example("slide123")
// @Success 200 {object} domain.SlideAnnotationsOverview "Slide annotations overview"
// @Failure 400 {object} domain.ErrorResponse "Bad request - slideUid required"
// @Failure 404 {object} domain.ErrorResponse "Slide not found"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides/{slideUid}/annotations [get]
func GetSlideAnnotationsOverview(slidesService services.SlidesService, masksService services.RasterAnnotationsService, vectorsService services.VectorAnnotationsService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var input SlideUIDParams
		if err := c.ParamsParser(&input); err != nil {
			log.Warn("GetSlideAnnotationsOverview request parsing failed", "error", err)
			return middleware.SendError(c, fiber.StatusBadRequest, "invalid request")
		}

		// Validate the parsed parameters
		if err := validation.GlobalValidator.ValidateStruct(c, input); err != nil {
			return err // Error already formatted and sent
		}

		// First verify that the slide exists
		_, err := slidesService.GetSlideByUID(c.UserContext(), input.SlideUID)
		if err != nil {
			log.Error("GetSlideAnnotationsOverview - slide lookup failed", "error", err, "slideUid", input.SlideUID)
			if errors.Is(err, apperrors.ErrSlideNotFound) {
				return middleware.SendError(c, fiber.StatusNotFound, "slide not found")
			}
			return middleware.HandleError(c, err)
		}

		// Get raster annotations for this slide
		rasterAnnotations, err := masksService.GetRasterAnnotationsForSlide(c.UserContext(), input.SlideUID)
		if err != nil {
			log.Error("GetSlideAnnotationsOverview - failed to get raster annotations", "error", err, "slideUid", input.SlideUID)
			return middleware.HandleError(c, err)
		}

		// Get vector annotations for this slide
		vectorAnnotations, err := vectorsService.GetVectorAnnotationsForSlide(c.UserContext(), input.SlideUID)
		if err != nil {
			log.Error("GetSlideAnnotationsOverview - failed to get vector annotations", "error", err, "slideUid", input.SlideUID)
			return middleware.HandleError(c, err)
		}

		// Build the response with URLs and counts
		overview := domain.SlideAnnotationsOverview{
			SlideUID:    input.SlideUID,
			RasterURL:   "/api/v1/slides/" + input.SlideUID + "/annotations/raster",
			VectorURL:   "/api/v1/slides/" + input.SlideUID + "/annotations/vector",
			RasterCount: len(rasterAnnotations),
			VectorCount: len(vectorAnnotations),
		}

		return c.JSON(overview)
	}
}

// UpdateSlide updates an existing slide using SlidesService
// @Summary Update a slide
// @Description Update an existing slide's information
// @Tags slides
// @Accept json
// @Produce json
// @Security BearerAuth
// @Param slideUid path string true "Slide UID" example("slide123")
// @Param slide body UpdateSlideInput true "Slide update information"
// @Success 200 {object} domain.Slide "Updated slide"
// @Failure 400 {object} domain.ErrorResponse "Bad request - invalid input"
// @Failure 404 {object} domain.ErrorResponse "Slide not found"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides/{slideUid} [put]
func UpdateSlide(slidesService services.SlidesService, imageTypesService services.ImageTypesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var input UpdateSlideInput
		if err := c.BodyParser(&input); err != nil {
			log.Warn("UpdateSlide request parsing failed", "error", err)
			return middleware.SendError(c, fiber.StatusBadRequest, "invalid request")
		}

		var params SlideUIDParams
		if err := c.ParamsParser(&params); err != nil {
			log.Warn("UpdateSlide request parsing failed", "error", err)
			return middleware.SendError(c, fiber.StatusBadRequest, "invalid parameters")
		}

		// Validate the parsed input
		if err := validation.GlobalValidator.ValidateStruct(c, input); err != nil {
			return err // Error already formatted and sent
		}

		// Validate the parsed parameters
		if err := validation.GlobalValidator.ValidateStruct(c, params); err != nil {
			return err // Error already formatted and sent
		}

		// Validate image type ID if provided in update
		if input.ImageTypeId != nil && *input.ImageTypeId != "" {
			_, err := imageTypesService.GetImageTypeByID(c.UserContext(), *input.ImageTypeId)
			if err != nil {
				log.Warn("Invalid image type ID provided in update", "imageTypeId", *input.ImageTypeId, "error", err)
				return middleware.SendError(c, fiber.StatusBadRequest, "invalid image type ID")
			}
		}

		// Get the existing slide
		existingSlide, err := slidesService.GetSlideByUID(c.UserContext(), params.SlideUID)
		if err != nil {
			log.Error("UpdateSlide - failed to get existing slide", "error", err, "slideUID", params.SlideUID)
			if errors.Is(err, apperrors.ErrSlideNotFound) {
				return middleware.SendError(c, fiber.StatusNotFound, "slide not found")
			}
			return middleware.HandleError(c, err)
		}

		// Apply updates to the existing slide
		updatedSlide := existingSlide
		if input.SlideName != nil {
			updatedSlide.SlideName = *input.SlideName
		}
		if input.SlideURI != nil {
			updatedSlide.SlideURI = *input.SlideURI
		}
		if input.ImageTypeId != nil {
			updatedSlide.ImageTypeId = *input.ImageTypeId
		}

		// Save the updated slide
		savedSlide, err := slidesService.SaveSlide(c.UserContext(), updatedSlide)
		if err != nil {
			log.Error("UpdateSlide failed", "error", err, "slideUID", params.SlideUID)
			if apperrors.IsInvalidInput(err) {
				return middleware.SendError(c, fiber.StatusBadRequest, err.Error())
			}
			return middleware.HandleError(c, err)
		}

		return c.JSON(savedSlide)
	}
}

// SoftDeleteSlide soft deletes a slide using SlidesService
// @Summary Soft delete a slide
// @Description Mark a slide as deleted without removing it from the database
// @Tags slides
// @Produce json
// @Security BearerAuth
// @Param slideUid path string true "Slide UID" example("slide123")
// @Success 200 {object} fiber.Map "Success message"
// @Failure 400 {object} domain.ErrorResponse "Bad request - slide ID required"
// @Failure 404 {object} domain.ErrorResponse "Slide not found"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides/{slideUid} [delete]
func SoftDeleteSlide(slidesService services.SlidesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var params SlideUIDParams
		if err := c.ParamsParser(&params); err != nil {
			log.Warn("SoftDeleteSlide request parsing failed", "error", err)
			return middleware.SendError(c, fiber.StatusBadRequest, "invalid parameters")
		}

		// Validate the parsed parameters
		if err := validation.GlobalValidator.ValidateStruct(c, params); err != nil {
			return err // Error already formatted and sent
		}

		// TODO: Get actual user ID from context
		err := slidesService.SoftDeleteSlide(c.UserContext(), params.SlideUID, 1)
		if err != nil {
			log.Error("SoftDeleteSlide failed", "error", err, "slideUID", params.SlideUID)
			if errors.Is(err, apperrors.ErrSlideNotFound) || errors.Is(err, apperrors.ErrSlideAlreadyDeleted) {
				return middleware.SendError(c, fiber.StatusNotFound, err.Error())
			}
			return middleware.SendError(c, fiber.StatusInternalServerError, "internal error")
		}

		return c.Status(fiber.StatusOK).JSON(fiber.Map{"message": "Slide deleted successfully"})
	}
}

// CreateSlide creates a new slide using SlidesService
// @Summary Create a new slide
// @Description Create a new slide in the system
// @Tags slides
// @Accept json
// @Produce json
// @Security BearerAuth
// @Param slide body SlideCreationInput true "Slide creation information"
// @Success 201 {object} domain.Slide "Created slide"
// @Failure 400 {object} domain.ErrorResponse "Bad request - invalid input"
// @Failure 500 {object} domain.ErrorResponse "Internal server error"
// @Router /api/v1/slides [post]
func CreateSlide(slidesService services.SlidesService, imageTypesService services.ImageTypesService) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var input SlideCreationInput
		if err := c.BodyParser(&input); err != nil {
			log.Warn("CreateSlide request parsing failed", "error", err)
			return middleware.SendError(c, fiber.StatusBadRequest, "invalid request")
		}

		// Validate the parsed input
		if err := validation.GlobalValidator.ValidateStruct(c, input); err != nil {
			return err // Error already formatted and sent
		}

		// Validate image type ID if provided
		if input.ImageTypeId != "" {
			_, err := imageTypesService.GetImageTypeByID(c.UserContext(), input.ImageTypeId)
			if err != nil {
				log.Warn("Invalid image type ID provided", "imageTypeId", input.ImageTypeId, "error", err)
				return middleware.SendError(c, fiber.StatusBadRequest, "invalid image type ID")
			}
		}

		slide := domain.Slide{
			SlideUID:    input.SlideUID,
			SlideName:   input.SlideName,
			SlideURI:    input.SlideURI,
			ImageTypeId: input.ImageTypeId,
		}

		createdSlide, err := slidesService.SaveSlide(c.UserContext(), slide)
		if err != nil {
			log.Error("CreateSlide failed", "error", err)
			if apperrors.IsInvalidInput(err) {
				return middleware.SendError(c, fiber.StatusBadRequest, err.Error())
			}
			return middleware.HandleError(c, err)
		}

		// Return 201 Created with the slide data
		c.Status(fiber.StatusCreated)
		return c.JSON(createdSlide)
	}
}
